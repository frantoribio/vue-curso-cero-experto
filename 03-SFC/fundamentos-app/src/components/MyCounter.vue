<template>
  <h2>Counter</h2>
  <p>{{ counter }} <sup>2</sup> = {{ counter * counter }}</p>
</template>

<script>
  export default {
    name: 'MyCounter',

    // Mi estado reactivo
    data: () => ({
      counter: 5,
    }),
  }
</script>

<style lang="scss" scoped></style>
