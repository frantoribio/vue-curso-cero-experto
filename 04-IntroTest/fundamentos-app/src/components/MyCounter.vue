<template>
  <h2>{{ titulo || 'alternativo' }}</h2>
  <p class="my-style">{{ counter }} <sup>2</sup> = {{ counterSquare }}</p>
  <div>
    <button
      data-testid="increment"
      @click="increment()"
    >
      +1
    </button>
    <button @click="decrement()">-1</button>
  </div>
</template>

<script>
  export default {
    name: 'MyCounter',

    // Mis propiedades
    props: {
      // count: Number, podemos definirla en una sola linea o con más opciones...
      titulo: {
        type: String, // Tipo de dato
        default: 'Contador', // valor por defecto (opcional)
        required: true, // es obligatorio pasarselo (opcional)
      },
      started: {
        type: Number,
        default: 0,
        required: true,
        // Validaro es opcional
        validator(value) {
          return value >= 2
        },
      },
    },

    // Mi estado reactivo, lo inicio según la propiedad!!
    data() {
      return {
        counter: this.started,
      }
    },

    // Mis métodos
    methods: {
      squareMethods() {
        return this.counter * this.counter
      },
      increment() {
        this.counter++
      },
      decrement() {
        this.counter--
      },
    },
    // Mis propiedades computadas, se gaurdan en cache y solo cambia cuando cambia el estado que computan..., solo son de lectura
    computed: {
      counterSquare() {
        return this.counter * this.counter
      },
    },
  }
</script>

<style scoped>
  h2 {
    color: blueviolet;
  }
  button {
    background-color: #64bb87;
    border-radius: 5px;
    color: white;
    margin: 0 5px;
    cursor: pointer;
    padding: 5px 10px;
    transition: 0.3s ease-in-out;
  }
  .my-style {
    color: #444444;
    font-size: 1.2em;
    font-weight: bold;
  }
</style>
