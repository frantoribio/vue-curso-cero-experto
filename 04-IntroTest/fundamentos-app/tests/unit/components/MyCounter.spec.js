import { describe, test, expect, beforeEach } from 'vitest'

import { shallowMount } from '@vue/test-utils'
import MyCounter from '@/components/MyCounter.vue'

// Suite para describir
describe('Component -> MyCounter', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(MyCounter)
  })

  // cada una de las pruebas o test
  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot() // snapshot es una imagen de la vista y podemos usar .element tambien o html()
  })

  test('renders properly', () => {
    // Montamos el componente y le pasamos las propiedades
    wrapper = shallowMount(MyCounter, { props: { titulo: 'Hello Vitest', started: '3' } })
    expect(wrapper.text()).toContain('Hello Vitest') // Comprobamos que el texto contiene el titulo
    expect(wrapper.text()).toContain('3') // Comprobamos que el texto contiene el valor
    expect(wrapper.text()).toContain('9') // Comprobamos que el texto contiene el valor del cuadrado
  })

  test('h2 debe tener un valor por defecto', () => {
    expect(wrapper.find('h2').exists()).toBeTruthy() // Comprobamos que el texto contiene el valor
    expect(wrapper.find('h2').text()).toBe('Contador')
  })

  test('los botones tienen las etiquetas que corresponden', () => {
    // encontramos todos los parrafos
    const [bIncrement, bDecrement] = wrapper.findAll('button')
    // expect(wrapper.findAll('button')).toBe(2) // Comprobamos que hay 2 botones
    expect(bIncrement.text()).toContain('+') // Comprobamos que el texto del boton es +
    expect(bDecrement.text()).toContain('-') // Comprobamos que el texto del boton es -
    const increment = wrapper.find('[data-testid="increment"]')
    expect(increment.exists()).toBeTruthy() // Comprobamos que el boton existe
    expect(increment.text()).toContain('+') // Comprobamos que el texto del boton es +
  })

  test('debe incrementar el valor al pulsar +1', async () => {
    const wrapper = shallowMount(MyCounter, { props: { titulo: 'Boton +1', started: '3' } })
    expect(wrapper.text()).toContain('3')
    expect(wrapper.text()).toContain('9')
    const bIncrement = wrapper.find('[data-testid="increment"]')
    await bIncrement.trigger('click')
    expect(wrapper.text()).toContain('4')
    expect(wrapper.text()).toContain('16')
  })

  test('debe decrementar el valor al pulsar -1', async () => {
    const wrapper = shallowMount(MyCounter, { props: { titulo: 'Boton -1', started: '3' } })
    expect(wrapper.text()).toContain('3')
    expect(wrapper.text()).toContain('9')
    const bIncrement = wrapper.find('[data-testid="decrement"]')
    await bIncrement.trigger('click')
    expect(wrapper.text()).toContain('2')
    expect(wrapper.text()).toContain('4')
  })

  test('Debemos poder leer el valor de las propiedades', () => {
    const wrapper = shallowMount(MyCounter, { props: { titulo: 'Titulo', started: '3' } })
    expect(wrapper.props().titulo).toBe('Titulo')
    expect(wrapper.props().started).toBe('3')
    expect(wrapper.find('h2').text()).toBe('Titulo')
     expect(wrapper.find('p').text()).contains('3')
  })
})
