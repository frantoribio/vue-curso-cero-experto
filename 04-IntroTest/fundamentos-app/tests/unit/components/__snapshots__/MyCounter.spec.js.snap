// Vitest Snapshot v1

exports[`Component -> MyCounter > Debe hacer match con el snapshot 1`] = `
"<h2 data-v-0352012b=\\"\\">Contador</h2>
<p class=\\"my-style\\" data-v-0352012b=\\"\\">0 <sup data-v-0352012b=\\"\\">2</sup> = 0</p>
<div data-v-0352012b=\\"\\"><button data-v-0352012b=\\"\\">+1</button><button data-v-0352012b=\\"\\">-1</button></div>"
`;
